File: server-config-examples/lucene/entity-defined-index/src/main/java/examples/springdata/geode/server/lucene/domain/Customer.java
package examples.springdata.geode.server.lucene.domain;

import examples.springdata.geode.domain.EmailAddress;
import org.springframework.data.annotation.Id;
import org.springframework.data.gemfire.mapping.annotation.LuceneIndexed;
import org.springframework.data.gemfire.mapping.annotation.PartitionRegion;

import java.io.Serializable;

@PartitionRegion(name = "Customers")
public class Customer implements Serializable {
    @Id
    private Long id;

    private EmailAddress emailAddress;

    private String firstName;

    @LuceneIndexed(name = "lastName_lucene")
    private String lastName;

    public Customer(long id, EmailAddress emailAddress, String firstName, String lastName) {
        this.id = id;
        this.emailAddress = emailAddress;
        this.firstName = firstName;
        this.lastName = lastName;
    }

    @Override
    public String toString() {
        return "Customer(id=" + id + ", emailAddress=" + emailAddress +", firstName=" + firstName + ", lastName=" + lastName + ")";
    }

    @Override
    public int hashCode() {
        return id == null ? 0 : id.hashCode();
    }
}

File: common-domain-models/src/main/kotlin/examples/springdata/geode/domain/Customer.kt
/*
 * Copyright 2012 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package examples.springdata.geode.domain

import org.springframework.data.annotation.Id
import org.springframework.data.gemfire.mapping.annotation.Region
import org.springframework.util.Assert
import java.io.Serializable
import javax.persistence.Entity

/**
 * A customer used for examples.
 *
 * @author Oliver Gierke
 * @author David Turanski
 * @author Udo Kohlmeyer
 */
@Entity
@Region(name = "Customers")
data class Customer (@Id @javax.persistence.Id val id: Long?, val emailAddress: EmailAddress,
                    val firstName: String, val lastName: String) : Serializable {

    private var addresses = mutableSetOf<Address>()

    init {
        Assert.hasText(firstName, "First Name cannot be empty")
        Assert.hasText(lastName, "Last Name cannot be empty")
    }

    /**
     * Adds the given [Address] to the [Customer].
     *
     */
    fun add(address: Address) = this.addresses.add(address)

    override fun hashCode(): Int = id?.hashCode() ?: 0
}
